### ISSUE TITLE
Scheduled arrival time for a stop is truncated

### ISSUE THREAD

**Issue by hypergreatthing** [OneBusAway/onebusaway-android#498]
Information displaying schedule arrival time is off by a minute in some cases.

I'm a little bit of an outsider in how the structure of this application is created, but i assume you're using the oba api to pull stop information that includes scheduled arrival times.

Stop level arrival information is merely estimated time between timepoints using distance.  The stop times are in seconds.

It's my understanding that this code: https://github.com/OneBusAway/onebusaway-android/blob/master/onebusaway-android/src/main/java/org/onebusaway/android/ui/ArrivalInfo.java
on line 228 is truncating the time values to minutes instead of rounding.
This causes a problem where a stop has the arrival time in gtfs of 3:42:53 but is showing up on the app as 3:42.

If it can only be displayed by the minute, at least the closest minute would be preferable.

**Comment by barbeau** [comment]
@hypergreatthing Good point!  Do you have a pointer to a specific region/route/stop where this is visible?  Would like to grab some data to create some unit tests.  Screenshot would help too.

I'm surprised this issue has never come up before, although maybe most agencies either round schedules to the minute or rounding down has been mostly accurate with agency data.

**Comment by barbeau** [comment]
This should also affect departure times too.

**Comment by barbeau** [comment]
And from a quick scan of the code on my phone I think real-time too.

**Comment by hypergreatthing** [comment]
WMATA: stop id 14078
I'm not sure if you use the siri call for stop info but this url will work:
http://buseta.wmata.com/api/stop-for-id?_=1463427413328&stopId=1_14078

I'll have to get a screenshot later.

**Comment by barbeau** [comment]
Out of curiosity - does this display correctly in the other OBA UIs (iPhone, web, etc.)?  Would help to have something accurate to test against.

**Comment by hypergreatthing** [comment]
![rosslyn](https://cloud.githubusercontent.com/assets/10711060/15324417/5e06e160-1c14-11e6-9658-00c37f0cad3a.jpg)
From GTFS:
868586830	13:49:57	13:49:57	14078	N MOORE ST + 19TH ST
868588830	14:29:57	14:29:57	14078	N MOORE ST + 19TH ST
Timetable:
http://www.wmata.com/bus/timetables/dc/05a.pdf?n

I don't think oba web shows scheduled arrival times when they have real time data.  When they do it's relative time (how many minutes until it should arrive).  I will ask however to make sure they're displaying the right data as it's also to the minute.

**Comment by hypergreatthing** [comment]
Just fyi they also truncate by design

**Comment by barbeau** [comment]
So WMATA is also truncating by design in their printed schedules?

Also, do you have a link to the full GTFS dataset being used?

**Comment by hypergreatthing** [comment]
No.  OBA is (as far as i know).
Basically stop times in GTFS are created through a algorithm which uses timepoints (scheduled times) overlayed on top of route shapes and uses the distances between the stops and shapes to determine what times the stops should be associated with.  In this particular instance the timepoint is 3 seconds (relative to distance) away from the timepoint.

There's the public GTFS on the WMATA site.  I can't five out the private GTFS that WMATA has given to be used in OBA unfortunately.

So the GTFS should have stops that are scheduled down to the second.

If you'd like i can give a sample of stop times for a particular stop.

**Comment by barbeau** [comment]
Can you provide link to WMATA GTFS you're looking at?

Even though non timepoint times are interpolated, the timepoint times
should still match the WMATA printed schedule.  If these aren't rounded to
the minute I'm curious how WMATA is displaying them, whether they are
truncating or rounding.

**Comment by hypergreatthing** [comment]
I can't provide a link as it's not public.  I will ask however.

The public gtfs however shows this same issue:
http://lrg.wmata.com/GTFS_data/google_transit.zip
stop #29136 in stop_times if you look at the almost all the arrival times is 6 seconds short of the minute.

I guess it's an issue with our system that we use.  Timepoints and stops have distinct locations that typically aren't the same physically.  While a lot of scheduling systems have a correlation between stops and timepoints, WMATA does not.

Because of that and that timepoints aren't in GTFS, stops that are typically associated to timepoints sometimes have different passing times by the order of seconds depending on the physical distance between the two.

**Comment by barbeau** [comment]
@hypergreatthing Thanks for all the info, I really appreciate it.  

Yes, this issue with truncating seconds off schedule and real-time values (and doing calculations based on minutes resolution) is a little more complicated than I initially thought, especially because there is no `timepoint` field in the WMATA [GTFS stop_times.txt file](https://developers.google.com/transit/gtfs/reference#stop_timestxt).  Technically this GTFS data violates the GTFS spec, as you're only supposed to provide non-timepoint arrivals/departures in stop_times.txt.  You can only provide interpolated times if you set `timepoint` to `0` for these records, and `timepoint` to `1` for timepoints.  Sounds like adding this field is challenging, though, because the timepoints don't correspond to physical stop locations.  I also wouldn't advise removing the interpolated data at this point - it's useful so at least the various apps consuming the GTFS data use the same interpolated schedule times (otherwise everyone could start using their own interpolation algorithms, which gets even more complicated).

My main concern with presenting this truncated info in OBA would be that we're calculating the ETA value (9 min) and the early/late status ("3 min late") using the truncated data - this could show +/- a minute based on seconds rounding errors (in OBA Android we currently always round down, which is the truncated value).  We could start calculating the ETA and status based on millisecond resolution, but unless we update the schedule time to stop truncating as well the ETA/schedule/current clock time isn't going to add up.  And arguably the display of clock schedule information presentation is ok as-is with the truncated values - normal digital clocks without a second display always round down.

Bottom line, I'm going to have to loop in the folks working on the other OBA UI interfaces to figure out how we want to handle this and be consistent across the various interfaces.  Not changing anything is obviously the easiest course of action, so we'll need to weight the pro/cons of trying to make any changes.  If this is a priority to WMATA please let us know.  In the mean-time I'm going to bump this issue to a future milestone.

cc @sheldonabrown for server/website UI and @aaronbrethorst for OBA iOS

**Comment by hypergreatthing** [comment]
A suggestion would be to round to the nearest minute in a fashion that isn't that much different then what you're doing now.
Ex: 
final long scheduledMins = scheduled / ms_in_mins;
changed to
final long scheduledMins = (scheduled+500) / ms_in_mins;

That would fix everything.  But i agree on being consistent throughout the different projects associated with OBA.

**Comment by barbeau** [comment]
>A suggestion would be to round to the nearest minute in a fashion that isn't that much different then what you're doing now.

I believe the main problem with this is that because we're calculating the ETA using minute precision, so you can end up with a schedule time, ETA, and clock time that don't match.  Currently, because we're always rounding down to closest minute, these always match.

For example, if the scheduled arrival time is `5:00:59 PM`, using the above calculation the scheduled time would be displayed to the user as `5:01 PM`.  If the bus is running on-time, and the device clock shows `5:00 PM`, the ETA would show as "NOW" (because we're basing ETA off the truncated `5:00:00 PM` time).  As a user I would expect this to show "1 min" - from my view it's currently 5:00 PM, and my bus isn't scheduled to arrive until 5:01 PM.

And, arguably rounding up for clock schedule time isn't behavior most people would expect.  My phone's clock in the status bar shows 5:00 PM until the seconds value hits 60, and then it increments to 5:01 PM.

A note for me - this is also quasi-related to clock sync issues - https://github.com/OneBusAway/onebusaway-android/issues/29 - as clock sync would affect the level of resolution (seconds) for ETAs/status that this issue is trying to achieve.

**Comment by hypergreatthing** [comment]
Ok.
Just fyi, i don't think OBA web shows scheduled arrival time as you do in the app.  When real-time isn't available they show "X minutes (scheduled)".  Your app by default shows the scheduled time of arrival truncated to the minute which makes it a lot more apparent.
A quick scan shows nyc, cta has to the second data in gtfs while mbta, la seems to round to the nearest minute.  It doesn't seem like a real standard has been set when it comes to arrival times.

**Comment by barbeau** [comment]
@hypergreatthing Thanks for that info.  Yes, I think there is a lot of variety here in how the data is represented - I'm sure this won't be the last time this topic comes up :).

**Comment by barbeau** [comment]
On a related note - I was browsing the Git history looking for something else, and ran across these two related issues:

* https://github.com/OneBusAway/onebusaway-android/issues/363
* https://github.com/OneBusAway/onebusaway-android/issues/375

We were previously calculating status colors (red, blue, green) in milliseconds precision, and had problems because these colors didn't always match the status text (On-time, delayed, etc.), because the status text was calculated in minutes.  So we switched to using minutes everywhere so all the values would match.  So, we don't want to go backwards and increase precision just to see these same issues again - if we switch, it needs to be a comprehensive change through the entire app, and likely on other platforms as well.
